### ISSUE TITLE
Precision of each word segmentation engine.

### ISSUE THREAD

**Issue by Zayajung** [PyThaiNLP/pythainlp#62]
รบกวนสอบถามความแม่นยำของแต่ละ engine (6 engine) ในการทำ thai tokenize สำหรับ deepcut ทราบตัวเลขแล้ว แต่ตัวอื่นๆไม่ทราบว่าหาได้จากไหนบ้างคะ พยายามหาบทความแล้ว ไม่เจอเลยค่ะ

เพื่อจะใช้ในการอ้างอิงในการเลือก engine ในการตัดคำ

ขอบคุณค่ะ

**Comment by spicydog** [comment]
ผมไม่แน่ใจว่าเคยมีใครเอา Engine พวกนี้มารันเก็บผลบน Corpus เดียวกันแล้วตีพิมพ์หรือเปล่า (แต่เท่าที่ทราบคือไม่มีครับ)

หลาย Engines สร้างออกมาไม่ได้เป็นไปตามกระบวนการวิจัยที่รัดกุมครับ และตัวตัดคำแต่ละตัวก็ใช้วิธีตัดที่แตกต่างกัน ซึ่ง Corpus ที่จะใช้ทำ Benchmark ก็มีวิธีการตัดคำที่แตกต่างกันเช่นคำ บางอันอิงคำมูล บางอันคำผสม บางอันตามความหมาย แต่ Corpus ที่นิยมใช้กันตอนนี้จะเป็น InterBest 2009 และ 2010 ครับ ส่วน Paper ที่อิงตาม Corpus นี้ก็มักจะไม่ได้มีรวมอยู่ใน Lib นี้ครับ ถ้าอยากได้ตัวเลขจริงอาจต้องโหลด Corpus แล้วเอา Engine ต่างๆ เข้าไปลองรันเก็บผลเองครับ แต่ผมคิดว่ามันจะไม่แฟร์ เพราะแต่ละ Engine ใช้กฎในการตัดคำที่แตกต่างกันครับ :)

**Comment by Zayajung** [comment]
ขอบคุณมากค่ะ

**Comment by bact** [comment]
มีบทความเปรียบเทียบลักษณะนี้อยู่เหมือนกันครับ แต่เป็นการเปรียบเทียบด้วยคลัง ORCHID (ตอนนั้นยังไม่มี BEST)
โดยเป็นการเปรียบเทียบการตัดคำแบบใช้พจนานุกรม (อัลกอริทึม longest matching และ maximal matching กับ พจนานุกรม Lexitron และ พจนานุกรมที่มีคำตามโดเมน) กับการตัดคำโดยใช้การเรียนรู้ของเครื่อง (naive bayes, decision tree [J48], support vector machine, และ conditional random field)

A comparative study on Thai word segmentation approaches (Haruechaiyasak 2008)
https://www.researchgate.net/publication/4362221_A_comparative_study_on_Thai_word_segmentation_approaches

**Comment by wannaphongcom** [comment]
อันนี้เป็นของ deepcut ครับ https://docs.google.com/spreadsheets/d/1KXyyycO9aVLkh5IDe1BTnYoO5Oevr0o1Gr3GvWauSZk/edit?usp=sharing โดย test กับ best testset ครับ

**Comment by wannaphongcom** [comment]
ส่วนของ newmm โดย test กับ best testset ได้ผลดังนี้ครับ

```
F-measure: 71.80441
Recall: 86725.00000/125850.00000 = 68.91140

Precision: 86725.00000/115709.00000 = 74.95096

Number of incorrect : 39125.00000 words
```

**Comment by wannaphongcom** [comment]
ส่วนของ pyicu โดย test กับ best testset ได้ผลดังนี้ครับ

```
F-measure: 29.31907
Recall: 40213.00000/125850.00000 = 31.95312

Precision: 40213.00000/148463.00000 = 27.08621

Number of incorrect : 85637.00000 words
```

**Comment by wannaphongcom** [comment]
สาเหตุที่ประสิทธิภาพต่างกันแม้จะ test โดยใช้ testset เดียวกัน ตามที่คุณ @spicydog กล่าวไว้ข้างบนครับ

**Comment by heytitle** [comment]
ทาง pythainlp มีแผนทำ benchmark collection ของ algorithms ต่างๆ ไหมครับ 

จริงๆ เห็นมี อันนี้อยู่ https://github.com/kobkrit/nlp_thai_resources แต่พอหน้ามันเริ่มใหญ่แล้วมันดูยาก

**Comment by c4n** [comment]
> ทาง pythainlp มีแผนทำ benchmark collection ของ algorithms ต่างๆ ไหมครับ
> 
> จริงๆ เห็นมี อันนี้อยู่ https://github.com/kobkrit/nlp_thai_resources แต่พอหน้ามันเริ่มใหญ่แล้วมันดูยาก

ที่จริงผมชอบ format แบบนี้มากกว่านะ(https://github.com/sebastianruder/NLP-progress)  อยากลองทำกันไหม @wannaphongcom  @heytitle  ทำเสร็จแล้วก็ส่งไปใส่หน้าของ ruder อีกรอบก็ได้

**Comment by heytitle** [comment]
จริงๆ อยากทำประมาณนี้เลยอ่ะ แบบ มี tracking performance ของแต่ละ algorithm ด้วย

https://paperswithcode.com

**Comment by cstorm125** [comment]
สำหรับ classification รวบรวมไว้ที่นี่ครับ
https://github.com/pythainlp/classification-benchmarks

ส่วน nlp-progress ผมคิดว่าเราอาจจะ PR ไปหาเขาก็ได้ (อย่างคนที่ทำ hindi2vec ก็ PR ไป) แต่ตอนนี้ kobkrit/nlp_thai_resources มันใหญ่แล้วเหมือนจะติดตลาดไปแล้วเลยยังไม่มีแผนครับ

**Comment by c4n** [comment]
@heytitle @wannaphongcom @spicydog @bact @Zayajung @cstorm125 

1. จริงๆตอนนี้ชุดข้อมูลเปิดสำหรับภาษาไทยมี 
https://www.nectec.or.th/corpus/index.php?league=pm  กับของtltk 
 รู้สึกว่าถ้าจะ evaluate  มันมีปัญหาตรงที่ว่าtrain-test-splitแต่ละคนแบ่งเอง แล้วก็แทบไม่เห็นใครทำcross-validation ซึ่งจริงๆ ถ้าทำแบบชิวๆก็พอประมาณผลได้ แต่ถ้าจะเทียบกับหลายๆคน มันจะเริ่มไม่แน่ใจเพราะอาจมีใครชนะเพราะได้ split ที่ง่ายก็ได้555 อย่างน้อยเราควรจะ report ไหมว่าแบ่งยังไง  หรือเราควรจะมีcommon testset?  (ส่วนพวกdownstream taskแบบWongnai, Wisesight  มันทำบน kaggle และมีชุด test แยกให้ อันนี้ไม่น่าลำบาก)  ถ้าใครแนะนำได้ก็บอกด้วยครับ

2. ตอนนี้ที่อยากทำให้pythainlpคือcode ส่วน preprocess data ของชุดข้อมูลเปิดพวกนี้  @wannaphongcom คิดว่าเหมาะกับโปรเจคนี้ไหม?

3.  ผมมี access ชุดข้อมูล proprietary ที่มีPOS/NER แน่นอนอัพโหลดลงในนี้ไม่ได้ แต่คิดว่าอาจจะเอา weight ที่ train แล้วลงมาได้(จะไปขออนุญาตเจ้าของข้อมูลครับ) ถ้าจะทำคิดว่า Deep Learning Framework (Keras,Pytorch, Allennlp,etc.) ไหนเหมาะที่จะเป็น backend ของ pythainlp ครับ  อยากจะเลือกมาอันนึง คิดว่าถ้าต่อไปต่างคนต่างทำแล้วแบบต้องลงLibraryทุกค่ายถึงจะใช้ได้อันนี้คงหนักอยู่555

4. kobkrit/nlp_thai_resources มันมีปัญหาเรื่อง formatting แบบที่ทุกคนพูดถึงเหมือนกันครับ มันใหญ่และดูยาก เห็นด้วยกับ @cstorm125 เรื่อง ส่งPRโดยตรงไปเลยก็ดีครับ

**Comment by cstorm125** [comment]
@c4n 
1. ส่วนของ classification อย่างที่บอกมี wisesight กับ wongnai ที่จารย์เอกพลแบ่ง test set ไปแล้ว อีกสองอันที่เหลือ (prachathai-67k / truevoice-intent) ผมก็ใช้วิธี random แบ่งไปเหมือนกันครับ อาจจะไม่ fool-proof เท่าแต่เวลาใครทำ benchmark ก็ให้ส่ง code มาด้วยน่าจะเชื่อถือได้ส่วนนึง ส่วนของ NECTEC BEST มี test set ที่ต้องส่งไปให้เขาตรวจ มันไม่ค่อยสะดวกแต่น่าจะเปรียบเทียบได้ corpus อื่นต้องรอผู้รู้

3. เห็นด้วยครับ Dependency ของ DL ตอนนี้มี tensorflow สำหรับ deepcut กับ pytorch เพื่อ ulmfit ส่วนตัวผมใช้ pytorch แต่คิดว่าอาจจะต้อง migrate ไป tensorflow เพื่อ performance ในบาง task เช่น ใช้ Go/Swift ทำ deepcut ที่มันเร็วพอใช้งานจริงได้ (ตอนนี้ตัด BEST 1 ชม. เทียบกับ dict ที่หน่วยเป็นวินาที) ซึ่งถึงเวลานั้นเราอาจจะต้องหาวิธีเรียกใช้โมเดลด้วย python อีกที อยากลอง scope functionality ที่จำเป็นต้องใช้ DL ออกมาให้หมดก่อนแล้วค่อย migrate ไป framework ใน framework หนึ่งครับ

**Comment by wannaphongcom** [comment]
Deep Learning Framework ควรเลือกตามความเหมาะสมครับ โดยตอนนี้ pythainlp ใช้ 2 อัน tensorflow สำหรับ Keras (ตัวแปลงเสียงคำไทยเป็นคำอังกฤษ) , deepcut  และ PyTorch สำหรับ Fastai (ใช้กับ ulmfit) ถ้าตามความคิดเห็นผม ผมคิดว่าควรใช้ pytorch เพราะโค้ดง่ายต่อการบำรุงรักษากว่า tensorflow และ tensorflow 2.0 ยังไม่ออกด้วยครับ

**Comment by c4n** [comment]
> Deep Learning Framework ควรเลือกตามความเหมาะสมครับ โดยตอนนี้ pythainlp ใช้ 2 อัน tensorflow สำหรับ Keras (ตัวแปลงเสียงคำไทยเป็นคำอังกฤษ) , deepcut และ PyTorch สำหรับ Fastai (ใช้กับ ulmfit) ถ้าตามความคิดเห็นผม ผมคิดว่าควรใช้ pytorch เพราะโค้ดง่ายต่อการบำรุงรักษากว่า tensorflow และ tensorflow 2.0 ยังไม่ออกด้วยครับ

ถ้าให้เลือก งั้นคงเป็น pytorch ครับ เพราะผมถนัดอันนี้กว่าtf ด้วย

**Comment by wannaphongcom** [comment]
ผลการทดสอบตัดคำ newmm โดย test กับ best testset ของ PyThaiNLP 2.0 ที่อยู่ใน GitHub ตอนนี้ได้ผลดังนี้ครับ

```
F-measure: 75.36646
Recall: 88149.00000/125850.00000 = 70.04291

Precision: 88149.00000/108071.00000 = 81.56582

Number of incorrect : 37701.00000 words
```

เพิ่มขึ้นจากเดิมที่ได้ 71% จากข้างบน

**Comment by heytitle** [comment]
@wannaphongcom  @cstorm125 

พอดีผมกำลังจะทำโปรเจ็คเกี่ยวกับ tokenisation เลยอยากสอบถามว่า ทาง PyThaiNLP มี module สำหรับ benchmark results ไหมครับ  เพราะเห็น algorithms เช่น DeepCut หรือ Sertis's [Thai Word Segmentation][1] คิด f-score จาก character level. ซึ่งอาจจะทำให้ค่าสูงเกินจริง (https://github.com/sertiscorp/thai-word-segmentation/issues/9). 

ถ้าไม่มี เราอาจจะช่วย dev ตัว module นี้

[1]: https://github.com/sertiscorp/thai-word-segmentation/

**Comment by wannaphongcom** [comment]
@heytitle ตอนนี้ยังไม่มีครับ

**Comment by heytitle** [comment]
@wannaphongcom  พอทราบไหมครับ ว่า test set ของ BEST  นี้โหลดได้จากที่ไหนหรอครับ?

**Comment by heytitle** [comment]
@wannaphongcom @c4n @bact 

ตัว benchmark ของ tokenisation  เวอร์ชั่นแรก ขึ้นแล้วน่ะครับ
https://github.com/heytitle/thai-word-tokenisation-benchmark

รบกวนขอความเห็นว่า ด้วยครับ โดยเฉพาะตรงตัว metrics อยากรู้ที่ทำไว้โอเคไหม และต้องมีอะไรเพิ่มเติมบ้าง

**Comment by wannaphongcom** [comment]
@heytitle testset http://nectec.or.th/corpus/

**Comment by cstorm125** [comment]
Should we close this and continue discussion at @heytitle 's repo?

**Comment by bact** [comment]
> @wannaphongcom @c4n @bact
> 
> ตัว benchmark ของ tokenisation เวอร์ชั่นแรก ขึ้นแล้วน่ะครับ
> https://github.com/heytitle/thai-word-tokenisation-benchmark
> 
> รบกวนขอความเห็นว่า ด้วยครับ โดยเฉพาะตรงตัว metrics อยากรู้ที่ทำไว้โอเคไหม และต้องมีอะไรเพิ่มเติมบ้าง

ถ้ามีอีก metric ที่บอกว่า tokenizer ตัวนี้ destructive หรือไม่/แค่ไหน (ตัวอักษรทุกตัวยังอยู่ครบ สามารถเอาทุกตัวอักษรกลับมาประกอบกันเป็นข้อความต้นฉบับได้ใกล้เคียงแค่ไหน) จะมีประโยชน์ไหมครับ
หรือคิดว่าตัวเลขจาก metric อื่นๆ ก็น่าจะสะท้อนอยู่แล้ว ว่ามัน destructive แค่ไหน

**Comment by bact** [comment]
งั้นขออนุญาตปิดตรงนี้ แล้วไปคุยกันที่ https://github.com/heytitle/thai-word-tokenisation-benchmark/issues/8 นะครับ
